app/actions/dat-middleware.js

```diff
@@ -0,0 +1,423 @@
+'use strict'
+
+import Dat from 'dat-node'
+import { encode } from 'dat-encoding'
+import fs from 'fs'
+import { basename, join as joinPath } from 'path'
+import { ipcRenderer, shell } from 'electron'
+import mkdirp from 'mkdirp-promise'
+import mirror from 'mirror-folder'
+import promisify from 'util-promisify'
+
+const readFile = promisify(fs.readFile)
+const writeFile = promisify(fs.writeFile)
+const { Notification } = window
+
+export default class DatMiddleware {
+  constructor ({ downloadsDir, dataDir }) {
+    this.downloadsDir = downloadsDir
+    this.dataDir = dataDir
+    this.dats = {}
+    this.listeners = []
+    this.execByType = {
+      UPDATE_TITLE: action => this.updateTitle(action),
+      REMOVE_DAT: action => this.removeDat(action),
+      TRY_ADD_DAT: action => this.tryAddDat(action),
+      REQUEST_DOWNLOAD: action => this.validateDownloadRequest(action),
+      TOGGLE_PAUSE: action => this.togglePause(action),
+      DOWNLOAD_SPARSE_DAT: action => this.downloadSparseDat(action),
+      CANCEL_DOWNLOAD_DAT: action => this.cancelDownloadDat(action)
+    }
+  }
+
+  async validateDownloadRequest ({ key }) {
+    if (key) {
+      key = encode(key)
+      if (this.dats[key]) {
+        return this.dispatch({ type: 'ADD_DAT_ERROR', key, error: new Error('Dat with same key already exists.') })
+      }
+    }
+    this.dispatch({ type: 'SHOW_DOWNLOAD_SCREEN', key})
+  }
+
+  execAction (action) {
+    const exec = this.execByType[action.type]
+    if (!exec) return false
+    // Telling it to all the middlewares.
+    this.dispatch(action)
+    exec(action)
+      .then(data =>
+        this.dispatch({ ...action, type: `${action.type}_SUCCESS`, ...data })
+      )
+      .catch(error =>
+        this.dispatch({ ...action, type: `${action.type}_ERROR`, error })
+      )
+    return true
```

Comment:
We can use `async/await` here.

Reply:
The `return true` statement returns before the `.then().catch()` is finished. As such the following [`dispatch`](https://github.com/dat-land/dat-desktop/pull/516/files#diff-66dae0682d14f11364314215e58de09dR67) is called in parallel to the execution.